core/src/main/scala/com/microsoft/azure/synapse/ml/causal/DoubleMLEstimator.scala

```diff
@@ -313,13 +359,151 @@ class DoubleMLModel(val uid: String)
   @Experimental
   override def transform(dataset: Dataset[_]): DataFrame = {
     logTransform[DataFrame]({
-      dataset.toDF()
+      val df = {
+        if (getTreatmentType == TreatmentTypes.Continuous) {
+          println("Warning: transform function will not calculate " +
+            "individual treatment effect for continuous treatment.")
+
+          dataset.withColumn(getIteOutputCol, lit(Double.NaN))
+            .withColumn(getIteStddevOutputCol, lit(Double.NaN))
+        }
+        else {
+          val treatmentModels = getTrainedTreatmentModels()
+          val outcomeModels = getTrainedOutcomeModels()
+          val regressorModels = getTrainedRegressorModels()
+          val idxColName = DatasetExtensions.findUnusedColumnName("idx", dataset)
+          val iteColName = DatasetExtensions.findUnusedColumnName("ite", dataset)
+
+          val dfIdx = dataset.withColumn(idxColName, monotonically_increasing_id())
+
+          val executionContext = getExecutionContextProxy
+          val transformDFFutures = (0 until  treatmentModels.length).toArray.map { index =>
+            Future[DataFrame] {
+              val processedDF = calculateIteCol(
+                dfIdx,
+                iteColName,
+                treatmentModels(index),
+                outcomeModels(index),
+                regressorModels(index))
+              processedDF
+            }(executionContext)
+          }
+
+          val dfs = awaitFutures(transformDFFutures)
+
+          val dfIte =
+            dfs.reduce((df1, df2) => df1.union(df2))
+              .groupBy(idxColName)
+              .agg(avg(iteColName).alias(getIteOutputCol),
+                stddev(iteColName).alias(getIteStddevOutputCol))
+
+          val dfRes =
+            dfIdx.join(dfIte, Seq(idxColName))
+              .select(dfIdx.columns.map(col) ++ Seq(col(getIteOutputCol), col(getIteStddevOutputCol)): _*)
+              .drop(idxColName)
+
+          dfRes
+        }
+
+      }
+      df
     })
   }
 
+  private def calculateIteCol(dataset: Dataset[_],
+                              iteColName: String,
+                              treatmentModel: Transformer,
+                              outcomeModel: Transformer,
+                              regressorModel: Transformer): DataFrame = {
```

Comment:
If you combine these three transformers into a pipeline model, you can still retrieve each one from the "stages" field, but will simply the interface.